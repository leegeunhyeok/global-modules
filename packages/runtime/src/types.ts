export interface Module {
  (exports: ModuleExports, require: ModuleRequire): void;
  id: ModuleId;
  exports: Exports;
  deps: DependencyMap;
  status: ModuleStatus;
}

export type ModuleStatus = 'idle' | 'stale' | 'ready';
export type ModuleExports = (definitions: Exports) => void;
export type ModuleRequire = (source: string) => ModuleExports;
export type Exports = Record<string, unknown>;

export type DependencyMap = Record<
  string,
  ModuleId | Exports | (() => Exports)
>;

export interface GlobalModuleRegistry {
  /**
   * Defines a global module.
   *
   * This function can handle two cases:
   * - Case 1: When defining the initial module.
   *   - Dependencies must be passed as an object(ESM) or a function(CommonJS).
   *   - Directly references the module object.
   * - Case 2: When redefining an existing module. (eg. HMR)
   *   - The ID of the defined module must be passed.
   *   - References the module corresponding to the ID.
   *
   * **Case 1**
   *
   * ```ts
   * define((exports, require) => {
   *   const mod0 = require('./foo');
   *   const mod1 = require('./bar');
   *   const mod2 = require('./baz');
   *
   *   // module body
   *
   *   exports({
   *     named: expr_1,
   *     default: expr_2,
   *   });
   * }, 1000, {
   *   './foo': foo,
   *   './bar': bar,
   *   './baz': () => baz, // when CommonJS or Dynamic imports (for lazy evaluation)
   * });
   * ```
   *
   * ---
   *
   * **Case 2**
   *
   * ```ts
   * define(..., 1001, {}); // foo
   * define(..., 1002, {}); // bar
   * define(..., 1003, {}); // baz
   *
   * define((exports, require) => {
   *   const mod0 = require('./foo');
   *   const mod1 = require('./bar');
   *   const mod2 = require('./baz');
   *
   *   // module body
   *
   *   exports({
   *     named: expr_1,
   *     default: expr_2,
   *   });
   * }, 1000, { './foo': 1001, './bar': 1002, './baz': 1003 });
   * ```
   *
   */
  define: (
    factory: (exports: ModuleExports, require: ModuleRequire) => void,
    id: ModuleId,
    deps?: DependencyMap,
    evaluate?: boolean,
  ) => void;
  /**
   * Re-evaluates the specified defined module and creates a new exports object.
   *
   * It also propagates the changes to ensure that inverse dependency modules(parents) reference the new exports object.
   *
   * ```ts
   * // Re-evaluates `1001` module with reference dependencies via id.
   * update(1010, { './foo': 1007, './bar': 1004, './baz': 1003 });
   * ```
   */
  update: (id: ModuleId, deps?: DependencyMap, evaluate?: boolean) => void;
  /**
   * Get module's status.
   */
  status: (id: ModuleId) => ModuleStatus;
  /**
   * Clear all modules from the registry.
   */
  clear: () => void;
}
