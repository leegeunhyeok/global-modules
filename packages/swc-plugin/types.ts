export enum Phase {
  Register = 0,
  Runtime = 1,
}

export interface PluginConfig {
  id: number;
  phase: Phase;
  dependencyIds?: Record<string, number>;
}
