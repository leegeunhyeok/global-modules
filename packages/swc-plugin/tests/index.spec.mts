import assert from 'node:assert';
import * as path from 'node:path';
import * as vm from 'node:vm';
import * as swc from '@swc/core';
import * as esbuild from 'esbuild';
import { vi, describe, it, expect } from 'vitest';
import { Phase } from '../types.js';

async function bundle(
  entryCode: string,
  code: string,
  { phase }: { phase: Phase },
) {
  const result = await swc.transformSync(code, {
    jsc: {
      target: 'es5',
      experimental: {
        plugins: [
          [
            path.resolve(
              import.meta.dirname,
              '../swc_plugin_global_modules.wasm',
            ),
            { id: '0', phase },
          ],
        ],
      },
    },
  });

  const buildResult = await esbuild.build({
    absWorkingDir: '/',
    bundle: true,
    write: false,
    format: 'iife',
    stdin: {
      contents: entryCode,
      loader: 'js',
    },
    plugins: [
      {
        name: 'linker',
        setup(build) {
          build.onResolve({ filter: /.*/ }, async () => ({
            path: '@',
            namespace: 'linker',
          }));

          build.onLoad({ filter: /.*/, namespace: 'linker' }, async () => ({
            contents: result.code,
            loader: 'js',
          }));
        },
      },
    ],
  });

  const bundleCode = buildResult.outputFiles[0]?.text;
  assert(bundleCode, 'invalid bundle result');

  return bundleCode;
}

function evaluate(code: string, context?: vm.Context) {
  const modules = {};

  function register(id) {
    if (modules[id]) {
      return { ...modules[id], reset: vi.fn() };
    }

    const module = { exports: {} };

    modules[id] = { module };

    return {
      module,
      exports: (factory) => {
        modules[id].module.exports = factory();
      },
      reset: vi.fn(),
    };
  }

  const runtimeContext = vm.createContext({
    global: { __modules: { register, getContext: register } },
    ...context,
  });

  return new vm.Script(code).runInContext(runtimeContext);
}

describe('@global-modules/swc-plugin', () => {
  describe('Bundle phase', () => {
    it('[ESM] Basics', async () => {
      const bundleCode = await bundle(
        `
        import * as mod from '.';
      
        bridge(mod);
        `,
        `
        const foo = 'foo';
        const bar = 'bar';
        const value = 'baz';

        export default 1;
        export const foo = 'foo';
        export { bar, value as baz };
        `,
        { phase: Phase.Bundle },
      );

      const bridge = vi.fn();
      evaluate(bundleCode, { bridge });

      expect(bundleCode).toMatchSnapshot();
      expect(bridge).toBeCalledWith({
        default: 1,
        foo: 'foo',
        bar: 'bar',
        baz: 'baz',
      });
    });

    it('[ESM] Export with declaration statements', async () => {
      const bundleCode = await bundle(
        `
        import * as mod from '.';

        bridge(mod.newObj);
        `,
        `
        const obj = { value: 0 };
        export const newObj = obj;

        newObj.key = 'key';
        `,
        { phase: Phase.Bundle },
      );

      const bridge = vi.fn();
      evaluate(bundleCode, { bridge });

      expect(bundleCode).toMatchSnapshot();
      expect(bridge).toBeCalledWith({ value: 0, key: 'key' });
    });

    it('[CJS] Basics', async () => {
      const bundleCode = await bundle(
        `
        const mod = require('.');

        bridge(mod);
        `,
        `
        const foo = 'foo';
        const bar = 'bar';
        const value = 'baz';

        if (true) {
          module.exports = { foo, bar, baz: value };
        }
        `,
        { phase: Phase.Bundle },
      );

      const bridge = vi.fn();
      evaluate(bundleCode, { bridge });

      expect(bundleCode).toMatchSnapshot();
      expect(bridge).toBeCalledWith({ foo: 'foo', bar: 'bar', baz: 'baz' });
    });
  });

  describe('Runtime phase', () => {
    it('[ESM] Basics', async () => {
      const bundleCode = await bundle(
        `
        import * as mod from '.';

        bridge(global.__modules.getContext(0).module.exports);
        `,
        `
        const foo = 'foo';
        const bar = 'bar';
        const value = 'baz';

        export default 1;
        export const foo = 'foo';
        export { bar, value as baz };
        `,
        { phase: Phase.Runtime },
      );

      const bridge = vi.fn();
      evaluate(bundleCode, { bridge });

      expect(bundleCode).toMatchSnapshot();
      expect(bridge).toBeCalledWith({
        default: 1,
        foo: 'foo',
        bar: 'bar',
        baz: 'baz',
      });
    });

    it('[ESM] Export with declaration statements', async () => {
      const bundleCode = await bundle(
        `
        import * as mod from '.';

        bridge(global.__modules.getContext(0).module.exports);
        `,
        `
        const obj = { value: 0 };
        export const newObj = obj;

        newObj.key = 'key';
        `,
        { phase: Phase.Runtime },
      );

      const bridge = vi.fn();
      evaluate(bundleCode, { bridge });

      expect(bundleCode).toMatchSnapshot();
      expect(bridge).toBeCalledWith({ newObj: { value: 0, key: 'key' } });
    });

    it('[CJS] Basics', async () => {
      const bundleCode = await bundle(
        `
        const mod = require('.');

        bridge(global.__modules.getContext(0).module.exports);
        `,
        `
        const foo = 'foo';
        const bar = 'bar';
        const value = 'baz';

        if (true) {
          module.exports = { foo, bar, baz: value };
        }
        `,
        { phase: Phase.Runtime },
      );

      const bridge = vi.fn();
      evaluate(bundleCode, { bridge });

      expect(bundleCode).toMatchSnapshot();
      expect(bridge).toBeCalledWith({ foo: 'foo', bar: 'bar', baz: 'baz' });
    });
  });
});
