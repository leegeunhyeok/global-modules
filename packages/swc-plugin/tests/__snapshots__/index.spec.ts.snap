// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@global-modules/swc-plugin > Bundle phase > [CJS] Basics 1`] = `
"(() => {
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __commonJS = (cb, mod2) => function __require() {
    return mod2 || (0, cb[__getOwnPropNames(cb)[0]])((mod2 = { exports: {} }).exports, mod2), mod2.exports;
  };

  // ../../../../../../dummy
  var require_dummy = __commonJS({
    "../../../../../../dummy"(exports, module) {
      var __ctx = global.__modules.register(0);
      var foo = "foo";
      var bar = "bar";
      var value = "baz";
      if (true) {
        module.exports = __ctx.module.exports = {
          foo,
          bar,
          baz: value
        };
      }
    }
  });

  // <stdin>
  var mod = require_dummy();
  bridge(mod);
})();
"
`;

exports[`@global-modules/swc-plugin > Bundle phase > [ESM] Basics 1`] = `
"(() => {
  var __defProp = Object.defineProperty;
  var __export = (target, all) => {
    for (var name in all)
      __defProp(target, name, { get: all[name], enumerable: true });
  };

  // ../../../../../../dummy
  var dummy_exports = {};
  __export(dummy_exports, {
    bar: () => bar,
    baz: () => value,
    default: () => dummy_default,
    foo: () => __x1
  });
  var __ctx = global.__modules.register(0);
  var foo = "foo";
  var bar = "bar";
  var value = "baz";
  var foo = "foo";
  __x1 = foo;
  __ctx.exports(function() {
    return {
      "default": __x,
      "foo": __x1,
      "bar": bar,
      "baz": value
    };
  });
  var __x;
  var __x1;
  var dummy_default = __x = 1;

  // <stdin>
  bridge(dummy_exports);
})();
"
`;

exports[`@global-modules/swc-plugin > Bundle phase > [ESM] Export with declaration statements 1`] = `
"(() => {
  // ../../../../../../dummy
  var __ctx = global.__modules.register(0);
  var obj = {
    value: 0
  };
  var newObj = obj;
  newObj.key = "key";
  __x = newObj;
  __ctx.exports(function() {
    return {
      "newObj": __x
    };
  });
  var __x;

  // <stdin>
  bridge(__x);
})();
"
`;

exports[`@global-modules/swc-plugin > Runtime phase > [CJS] Basics 1`] = `
"(() => {
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __commonJS = (cb, mod2) => function __require() {
    return mod2 || (0, cb[__getOwnPropNames(cb)[0]])((mod2 = { exports: {} }).exports, mod2), mod2.exports;
  };

  // ../../../../../../dummy
  var require_dummy = __commonJS({
    "../../../../../../dummy"() {
      var __ctx = global.__modules.getContext(0);
      __ctx.reset();
      var foo = "foo";
      var bar = "bar";
      var value = "baz";
      if (true) {
        __ctx.module.exports = {
          foo,
          bar,
          baz: value
        };
      }
    }
  });

  // <stdin>
  var mod = require_dummy();
  bridge(mod);
})();
"
`;

exports[`@global-modules/swc-plugin > Runtime phase > [ESM] Basics 1`] = `
"(() => {
  var __create = Object.create;
  var __defProp = Object.defineProperty;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __getProtoOf = Object.getPrototypeOf;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __commonJS = (cb, mod2) => function __require() {
    return mod2 || (0, cb[__getOwnPropNames(cb)[0]])((mod2 = { exports: {} }).exports, mod2), mod2.exports;
  };
  var __copyProps = (to, from, except, desc) => {
    if (from && typeof from === "object" || typeof from === "function") {
      for (let key of __getOwnPropNames(from))
        if (!__hasOwnProp.call(to, key) && key !== except)
          __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
    }
    return to;
  };
  var __toESM = (mod2, isNodeMode, target) => (target = mod2 != null ? __create(__getProtoOf(mod2)) : {}, __copyProps(
    // If the importer is in node compatibility mode or this is not an ESM
    // file that has been converted to a CommonJS file using a Babel-
    // compatible transform (i.e. "__esModule" has not been set), then set
    // "default" to the CommonJS "module.exports" for node compatibility.
    isNodeMode || !mod2 || !mod2.__esModule ? __defProp(target, "default", { value: mod2, enumerable: true }) : target,
    mod2
  ));

  // ../../../../../../dummy
  var require_dummy = __commonJS({
    "../../../../../../dummy"() {
      var __ctx = global.__modules.getContext(0);
      __ctx.reset();
      var foo = "foo";
      var bar = "bar";
      var value = "baz";
      __x = 1;
      var foo = "foo";
      __x1 = foo;
      __ctx.exports(function() {
        return {
          "default": __x,
          "foo": __x1,
          "bar": bar,
          "baz": value
        };
      });
      var __x;
      var __x1;
    }
  });

  // <stdin>
  var mod = __toESM(require_dummy());
  bridge(global.__modules.getContext(0).module.exports);
})();
"
`;

exports[`@global-modules/swc-plugin > Runtime phase > [ESM] Export with declaration statements 1`] = `
"(() => {
  var __create = Object.create;
  var __defProp = Object.defineProperty;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __getProtoOf = Object.getPrototypeOf;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __commonJS = (cb, mod2) => function __require() {
    return mod2 || (0, cb[__getOwnPropNames(cb)[0]])((mod2 = { exports: {} }).exports, mod2), mod2.exports;
  };
  var __copyProps = (to, from, except, desc) => {
    if (from && typeof from === "object" || typeof from === "function") {
      for (let key of __getOwnPropNames(from))
        if (!__hasOwnProp.call(to, key) && key !== except)
          __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
    }
    return to;
  };
  var __toESM = (mod2, isNodeMode, target) => (target = mod2 != null ? __create(__getProtoOf(mod2)) : {}, __copyProps(
    // If the importer is in node compatibility mode or this is not an ESM
    // file that has been converted to a CommonJS file using a Babel-
    // compatible transform (i.e. "__esModule" has not been set), then set
    // "default" to the CommonJS "module.exports" for node compatibility.
    isNodeMode || !mod2 || !mod2.__esModule ? __defProp(target, "default", { value: mod2, enumerable: true }) : target,
    mod2
  ));

  // ../../../../../../dummy
  var require_dummy = __commonJS({
    "../../../../../../dummy"() {
      var __ctx = global.__modules.getContext(0);
      __ctx.reset();
      var obj = {
        value: 0
      };
      var newObj2 = obj;
      newObj2.key = "key";
      __x = newObj2;
      __ctx.exports(function() {
        return {
          "newObj": __x
        };
      });
      var __x;
    }
  });

  // <stdin>
  var mod = __toESM(require_dummy());
  bridge(void 0);
})();
"
`;
